### ISSUE TITLE
Provide game zooming functionality

### ISSUE THREAD

**Issue by theradeonxt** [Keriew/julius#16]
**How to use:**
- to be based on _**mouse wheel scrolling**_ (_optional with Ctrl pressed at the same time_)
- allowed only in the **_city view_** or map editor view (_optional extension to other UI elements_).
- the user is allowed between 50 – 150 percent, in steps of 10 which gives the user 10 total levels of zoom.

**Demo**
![poc](https://user-images.githubusercontent.com/13402104/75282921-8612e280-581a-11ea-9d0f-2d51fc144aad.gif)
Test versions:

- Windows 64bit base julius 1.3.1 (old)
https://drive.google.com/file/d/1t7W-UEHgnO-UwXs6UZhwY1huVLtU296M/view?usp=sharing
- Windows 64bit julius roadblocks Keriew (latest):
https://drive.google.com/file/d/1ia_cuNAJoN8EZOYi6D5OfaSp3ZgHG6kC/view?usp=sharing

**Limitations**
- Other UI elements are also scaled, which makes the game very hard to interact with when zooming out too much.
- The zoom is not saved after game restarts.
- Desktop versions of julius (no switch or vita)
- As @Keriew noted, to keep the UI unnaffected, it is not a long term solution, needing other changes in how the game is drawn.

**Proof of concept**
Using [ and ] keys. Subject to change.
Branch: 
https://github.com/theradeonxt/julius

**Impact and memory use**
|native width | native height	|	zoom	| virtual width | virtual height		| C3 Ram use |
| ------------- |---| ---| ---| ---| ---|
|2560	|1017	|50% zoom out	|5120.0	|2034	|**452 MB**|
|2560	|1017	|none	|2560.0	|1017	|146 MB|
|2560	|1017	|150 zoom in	|1706.0	|678	|110 MB|

A 50% zoom out produces a 4x increase in the rendered image, this is expected.

**Questions:**
- Deciding for a min and max zoom? i.e a 5000 * 2000 image may seem too big.
- Allow zoom for everything inside the game (empire, main menu) or just city?

**Comment by crudelios** [comment]
In lines `131´to `136´of `src/platform/screen.c`, you have:

```
    if (scale_percentage != 100) {
        // Scale using nearest neighbour when we scale a multiple of 100%: makes it look sharper
        SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, (scale_percentage % 100 == 0) ? "nearest" : "linear");
        SDL_RenderSetLogicalSize(SDL.renderer, logical_width, logical_height);
    }
```

If you get rid of the `if` and keep just what's inside. Meaning keep this:

```
        // Scale using nearest neighbour when we scale a multiple of 100%: makes it look sharper
        SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, (scale_percentage % 100 == 0) ? "nearest" : "linear");
        SDL_RenderSetLogicalSize(SDL.renderer, logical_width, logical_height);
```

You may fix the fullscreen bug. I have't tested though.

**Comment by theradeonxt** [comment]
I will check, good tip though!

**Comment by theradeonxt** [comment]
By keeping the SDL_RenderSetLogicalSize for either mode it makes things better - no more left/right not updated images!
What is left is double menu bar(plus a bottom region of a few px that keeps the old image):
![menu](https://user-images.githubusercontent.com/13402104/75287758-a5623d80-5823-11ea-9755-94fd10b93a44.png)

**Comment by crudelios** [comment]
Can you get me the new build, please?

**Comment by theradeonxt** [comment]
Sure, this is without the if:
https://drive.google.com/file/d/1lKfKeKXhuJv76kgK9AoXrpU6IHF7W465/view?usp=sharing

**Comment by crudelios** [comment]
I got the problem.

You can't simply use the `width` and `height`from the settings during fullscreen as those are not updated when the window is set to full screen. So when you get those values from the settings `change_zoom` you're effectively telling the game to use the latest windowed size in the config... Whatever that was.

To fix the fullscreen issue, I would suggest you separate the new texture creation from `platform_screen_resize` and add it to a new function. 

`platform_screen_resize` should call a `platform_rescale_texture` which either gets the `width` and `height` from the settings if `setting_fullscreen()` is `0`, or uses [`SDL_GetDesktopDisplayMode`](https://wiki.libsdl.org/SDL_GetDesktopDisplayMode) to get the `width` and `height` values when in fullscreen.

Then, from `change_zoom`, you directly call `platform_rescale_texture` without ever needing to specify the current width or height.

**Comment by theradeonxt** [comment]
Seems the change was a coincidence.
The window settings I used but making sure to not save to file. There are more places where global width height variables are stored. I hope i use the right ones.
I appreciate the feedback, and will continue like you suggested. Thanks!

**Comment by theradeonxt** [comment]
Fixed fullscreen problem using different way to get w and h.
Left zooming enabled for every view. For example if I zoom out while in city, then go to empire or advisors, it can be too small to use.

**Comment by Keriew** [comment]
That's a really nice addition. I think ctrl+mouse wheel would be better, because square bracket keys are fairly frequently used for their main functionality. @theradeonxt

**Comment by Keriew** [comment]
Implemented now with 7432767
